File: Biblioteca-de-Trabajo/3. CODIGO/Prototipo v1.3/src/negocio/ServicioCitas.java
package negocio;

import datos.CitaRepo;
import modelos.Cita;
import java.util.List;
import java.util.UUID;

public class ServicioCitas implements IServicioFiCitas {
    private CitaRepo citaRepo = new CitaRepo();
    private GestorEventos gestor;
    private int usuarioLogueadoId;

    public ServicioCitas(GestorEventos gestor, int usuarioLogueadoId) {
        this.gestor = gestor;
        this.usuarioLogueadoId = usuarioLogueadoId;
    }

    public static class OpResult {
        public boolean exito;
        public String mensaje;
        public Cita cita;
        public OpResult(boolean e, String m) { exito = e; mensaje = m; }
        public OpResult(boolean e, String m, Cita c) { exito = e; mensaje = m; cita = c; }
    }

    public List<Cita> obtenerTodas() {
        return citaRepo.obtenerTodas();
    }

    public OpResult registrarCita(int pacienteId, String servicio, String consultorio, String fecha, String hora, int duracion) {
        String codigo = "CITA-" + UUID.randomUUID().toString().substring(0, 5).toUpperCase();
        Cita nueva = new Cita(0, pacienteId, servicio, consultorio, fecha, hora, duracion, codigo, "Programada");
        citaRepo.crear(nueva);
        gestor.notificar(String.valueOf(usuarioLogueadoId), "Auxiliar", "CREAR_CITA", "Código: " + codigo);
        return new OpResult(true, "Cita agendada. Código: " + codigo, nueva);
    }

    public OpResult cancelarCita(int citaId, String motivo) {
        if(motivo == null || motivo.trim().isEmpty()) {
            return new OpResult(false, "El motivo de cancelación es obligatorio.");
        }
        Cita c = citaRepo.obtenerPorId(citaId);
        if (c == null) return new OpResult(false, "Cita no encontrada.");

        c.setEstado("Cancelada");
        c.setMotivoCancelacion(motivo);
        citaRepo.actualizar(c);
        gestor.notificar(String.valueOf(usuarioLogueadoId), "Auxiliar", "CANCELAR_CITA", "Motivo: " + motivo);
        return new OpResult(true, "Cita cancelada exitosamente.");
    }
}

File: Biblioteca-de-Trabajo/3. CODIGO/Prototipo v2.0/src/negocio/ServicioCitas.js
import { Cita } from '../modelos/Cita.js';
import { CitaRepo } from '../datos/CitaRepo.js';

export class ServicioCitas {
  constructor(gestor, usuarioId, citaRepo) {
    this.gestor = gestor;
    this.usuarioId = usuarioId;
    this.citaRepo = citaRepo || new CitaRepo();
  }

  async obtenerTodas() {
    return this.citaRepo.listar();
  }

  async registrarCita(pacienteId, servicio, consultorio, fecha, hora, duracion, fisioterapeuta) {
    const cita = new Cita({ id: Date.now(), pacienteId, servicio, consultorio, fecha, hora, duracion, fisioterapeuta, estado: 'Programada' });
    await this.citaRepo.guardar(cita);
    this.gestor.notificar({ tipo: 'CITA_REGISTRADA', datos: cita });
    return { exito: true, mensaje: 'Cita registrada', cita };
  }

  async cancelarCita(id, motivo) {
    const cita = await this.citaRepo.actualizar(id, { estado: 'Cancelada', motivoCancelacion: motivo });
    if (!cita) {
      return { exito: false, mensaje: 'Cita no encontrada' };
    }
    this.gestor.notificar({ tipo: 'CITA_CANCELADA', datos: cita });
    return { exito: true, mensaje: 'Cita cancelada', cita };
  }

  async reprogramarCita(id, fecha, hora, servicio, consultorio, fisioterapeuta) {
    const citaActual = await this.citaRepo.obtenerPorId(id);
    if (!citaActual) {
      return { exito: false, mensaje: 'Cita no encontrada' };
    }
    
    const update = { 
      fecha, 
      hora, 
      estado: 'Reprogramada',
      servicio: servicio || citaActual.servicio,
      consultorio: consultorio || citaActual.consultorio,
      fisioterapeuta: fisioterapeuta || citaActual.fisioterapeuta
    };

    const cita = await this.citaRepo.actualizar(id, update);
    this.gestor.notificar({ tipo: 'CITA_REPROGRAMADA', datos: cita });
    return { exito: true, mensaje: 'Cita reprogramada', cita };
  }
}
